cmds/create-ffs/create-ffs.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2019 the LinuxBoot Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"crypto/sha1"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/linuxboot/fiano/pkg/guid"
+	"github.com/linuxboot/fiano/pkg/uefi"
+	"github.com/linuxboot/fiano/pkg/visitors"
+)
+
+var (
+	debug      = flag.Bool("d", false, "Enable debug prints")
+	outfile    = flag.String("o", "", "output file, default is stdout")
+	name       = flag.String("name", "", "Name to include in UI section")
+	filetype   = flag.String("type", "DRIVER", "UEFI filetype")
+	version    = flag.String("version", "1.0", "File version")
+	guidString = flag.String("guid", "", "File GUID")
+	depex      = flag.String("depex", "", "Space-separated protocol guid dependencies or TRUE")
+	compress   = flag.Bool("compress", false, "Wrap section data in a compressed section")
+	auto       = flag.Bool("auto", false, "Attempt to determine section types from file extensions")
+
+	fType  uefi.FVFileType
+	fGUID  *guid.GUID
+	depOps []uefi.DepExOp
+	printf = func(string, ...interface{}) {}
+)
+
+const (
+	usageString = "Usage: create-ffs [flags] file.efi [...]"
+)
+
+func createDepExes(deps string) ([]uefi.DepExOp, error) {
+	var err error
+	ops := []uefi.DepExOp{}
+
+	if strings.ToUpper(deps) == "TRUE" {
+		// Create just "TRUE" and "END" for now, but this feels unnecessary to me.
+		ops = append(ops, uefi.DepExOp{OpCode: "TRUE"})
+		ops = append(ops, uefi.DepExOp{OpCode: "END"})
+		return ops, nil
+	}
+
+	// we expect a space-separated list of GUIDs
+	guids := strings.Split(deps, " ")
+	numGUIDS := len(guids)
+	for _, guidStr := range guids {
+		var g *guid.GUID
+
+		if g, err = guid.Parse(guidStr); err != nil {
+			return nil, err
+		}
+		printf("depex guid requested: %v", *g)
+		ops = append(ops, uefi.DepExOp{OpCode: "PUSH", GUID: g})
+	}
+
+	// Append an "AND" for n-1 pushes.
+	for i := 1; i < numGUIDS; i++ {
+		ops = append(ops, uefi.DepExOp{OpCode: "AND"})
+	}
+
+	ops = append(ops, uefi.DepExOp{OpCode: "END"})
+	return ops, nil
+}
+
+func parseFlags() error {
+	var ok bool
+	var err error
+
+	if *debug {
+		printf = log.Printf
+	}
+
+	// Check filetypes
+	fType, ok = uefi.NamesToFileType[*filetype]
+	if !ok {
+		validTypes := []string{}
+		for k := range uefi.NamesToFileType {
+			validTypes = append(validTypes, k)
+		}
+		return fmt.Errorf("unable to get EFI File type, got %v, expected values in\n{%v}",
+			*filetype, strings.Join(validTypes, ", "))
+	}
+
+	if *guidString != "" {
+		fGUID, err = guid.Parse(*guidString)
+		if err != nil {
+			return err
+		}
+	} else if *name != "" {
+		// We sha1 the name to get a reproducible GUID.
+		fGUID = &guid.GUID{}
+		sum := sha1.Sum([]byte(*name))
+		copy(fGUID[:], sum[:guid.Size])
```

Comment:
Does this create a valid GUID? Guids contain version numbers.

Can you use something like https://godoc.org/github.com/google/uuid#NewSHA1 ?

Reply:
Trammell did this, I don't think anyone actually cares if a GUID is a "valid GUID" anymore.

Also, why depend on the google uuid package?